src/metabase/query_processor/middleware/resolve_fields.clj

```diff
@@ -12,12 +12,17 @@
   (when-let [parent-ids (seq (filter some? (map (comp :parent_id qp.store/field) field-ids)))]
     (recur parent-ids)))
 
+(defn all-field-ids
+  "Given a query, finds the set of all field IDs it references."
+  [query]
+  (into (set (mbql.u/match (:query query) [:field (id :guard integer?) _] id))
+        (comp cat (keep :id))
+        (mbql.u/match (:query query) {:source-metadata source-metadata} source-metadata)))
```

Comment:
I can't find where this is being used

Reply:
But I would highly support using this instead of trying to duplicate the logic elsewhere.

Reply:
Ah, it was needed for a middle permutation of serialization, but I found a better way to do it that didn't need this step.

Reply:
I've inlined this back into `resolve-fields`.